File: Birthday.java
// javac Birthday.java && java Birthday
import java.util.concurrent.CountDownLatch;
public class Birthday {
	static final int DAYS_IN_YEAR = 365;
	static final int NUM_THREADS = 768;
	static final int PEOPLE = 24;
	static final int TOTAL_SIMULATIONS = 1_000_000;
	static final int SIMULATIONS_PER_THREAD = TOTAL_SIMULATIONS / NUM_THREADS;
	static final int MULTIPLIER = 1664525;
	static final int INCREMENT = 1013904223;
	static class Worker implements Runnable {
		private int threadId;
		private int[] dataSuccessCount;
		private CountDownLatch latch;
		Worker(int threadId, int[] dataSuccessCount, CountDownLatch latch) {
			this.threadId = threadId;
			this.dataSuccessCount = dataSuccessCount;
			this.latch = latch;
		}
		@Override
		public void run() {
			long seed = System.nanoTime();
			int state = (int) (seed ^ threadId);
			int localSuccessCount = 0;
			for (int sim = 0; sim < SIMULATIONS_PER_THREAD; sim++) {
				int[] birthdays = new int[DAYS_IN_YEAR];
				for (int i = 0; i < PEOPLE; i++) {
					state = (state * MULTIPLIER + INCREMENT);
					int birthday = Integer.remainderUnsigned(state, DAYS_IN_YEAR);
					birthdays[birthday]++;
				}
				int exactlyTwoCount = 0;
				for (int i = 0; i < DAYS_IN_YEAR; i++) {
					if (birthdays[i] == 2) {
						exactlyTwoCount++;
					}
				}
				if (exactlyTwoCount == 1) {
					localSuccessCount++;
				}
			}
			dataSuccessCount[threadId] = localSuccessCount;
			latch.countDown();
		}
	}
	public static void main(String[] args) {
		long startTime = System.nanoTime();
		int[] dataSuccessCount = new int[NUM_THREADS];
		CountDownLatch latch = new CountDownLatch(NUM_THREADS);
		for (int t = 0; t < NUM_THREADS; t++) {
			new Thread(new Worker(t, dataSuccessCount, latch)).start();
		}
		try {
			latch.await();
		} catch (InterruptedException e) {
			return;
		}
		int totalSuccessCount = 0;
		for (int t : dataSuccessCount) totalSuccessCount += t;
		double probability = (double) totalSuccessCount / TOTAL_SIMULATIONS;
		System.out.printf("Probability: %.9f%n", probability);
		long endTime = System.nanoTime();
		double elapsedTime = (endTime - startTime) / 1e9;
		System.out.printf("Execution Time: %.3f s%n", elapsedTime);
	}
}

File: Birthday.kt
// kotlinc Birthday.kt -include-runtime -d Birthday.jar && java -jar Birthday.jar
import kotlin.concurrent.thread

const val DAYS_IN_YEAR: Int = 365
const val NUM_THREADS: Int = 768
const val PEOPLE: Int = 24
const val TOTAL_SIMULATIONS: Int = 1_000_000
const val SIMULATIONS_PER_THREAD: Int = TOTAL_SIMULATIONS / NUM_THREADS
const val MULTIPLIER: Int = 1664525
const val INCREMENT: Int = 1013904223

fun simulate(threadId: Int, successCount: IntArray): Unit {
  val seed: Long = System.nanoTime()
  var state: Int = (seed xor threadId.toLong()).toInt()
  var localSuccessCount: Int = 0
  for (sim: Int in 0 until SIMULATIONS_PER_THREAD) {
    val birthdays: IntArray = IntArray(DAYS_IN_YEAR)
    for (i: Int in 0 until PEOPLE) {
      state = state * MULTIPLIER + INCREMENT
      val birthday: Int = (state.toUInt() % DAYS_IN_YEAR.toUInt()).toInt()
      birthdays[birthday]++
    }
    var exactlyTwoCount: Int = 0
    for (i: Int in 0 until DAYS_IN_YEAR) {
      if (birthdays[i] == 2) {
        exactlyTwoCount++
      }
    }
    if (exactlyTwoCount == 1) {
      localSuccessCount++
    }
  }
  successCount[threadId] = localSuccessCount
}

fun main(): Unit {
  val startTime: Long = System.nanoTime()
  val successCount: IntArray = IntArray(NUM_THREADS)
  val threads: Array<Thread?> = arrayOfNulls<Thread>(NUM_THREADS)
  for (t: Int in 0 until NUM_THREADS) threads[t] = thread { simulate(t, successCount) }
  for (t: Thread? in threads) t?.join()
  var totalSuccessCount: Int = 0
  for (t: Int in successCount) totalSuccessCount += t
  val probability: Double = totalSuccessCount.toDouble() / TOTAL_SIMULATIONS
  System.out.printf("Probability: %.9f\n", probability)
  val endTime: Long = System.nanoTime()
  val elapsedTime: Double = (endTime - startTime) / 1e9
  System.out.printf("Execution Time: %.3f s\n", elapsedTime)
}
